File: BQ-24261-Charger/c/charger.c
// Original instructions - /* compile with cc -o mma7660 mma7660.c -lncurses */
// My changes -- //gcc -o mma mma.c -lncurses -l wiringPi -lm

#define BQ24261_ADDR 0x6B 		// taken from datasheet - page 27

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <signal.h>
#include <math.h>
#include <fcntl.h>
#include <linux/i2c-dev.h>
#include <time.h>
#include <sys/time.h>
#include <curses.h>
#include <wiringPi.h>

typedef unsigned char byte;

int deviceDescriptor;

void I2cSendData(byte addr,byte *data,int len)
{
        if(ioctl(deviceDescriptor,I2C_SLAVE, addr))
                printf("I2cSendData_device : IOCTL Problem\n");

        write(deviceDescriptor,data,len);
}

void I2cReadData(byte addr,byte *data,int len)
{
        if(ioctl(deviceDescriptor,I2C_SLAVE, addr))
                printf("I2cReadData_device : IOCTL Problem\n");

        read(deviceDescriptor,data,len);
}

// Inititalizes the I2C channel

void init_i2c(char *DeviceName)
{	
	printf("Initialising i2c device");

        deviceDescriptor=open(DeviceName, O_RDWR);

        if (deviceDescriptor == -1)
        {
                printf("Error opening device '%s'\n",DeviceName);
                exit(-1);
        }
}

// Entry point
int main(int argc, char **argv)
{
	/* Variable Declaration */

        int i;
	double v, voltage, timer, parts[2];
  	byte data[20], data0[2], data1[2], data2[2], data3[2], data4[2], data5[2], data6[2];
	FILE *file1;
	const char *filename1 = "testdata.txt";
	unsigned char file_data[100];
	time_t rawtime;
	struct tm *timeinfo;
	
	/* Execution */

	printf("Inside main \n");
	
	// the charger is connected to I2C bus-1 on RPi
        init_i2c("/dev/i2c-1");
	
	/* Set the registers of charger */

	// first assignment tells the address - total 7 registers (0-6)
	// the next assignment gives the actual value to be stored in decimal format
	// To understand the value -- convert to binary - pad zeros on the left to make 8 bits -
	//	                      each bit represents value for corresponding regitser bit B0 - B7
     
	data0[0] = 0;
	data0[1] = 148; // 16; //144;   // make the voltage permanent
	data1[0] = 1;
	data1[1] = 64;
	data2[0] = 2;
	data2[1] = 4;
	data3[0] = 3;
	data3[1] = 70;
	data4[0] = 4;
	data4[1] = 42;
	data5[0] = 5;
	data5[1] = 0;
	data6[0] = 6;
	data6[1] = 112;
	
	// Open the file in read-only mode
	file1 = fopen(filename1, "rb");
	
	if(file1 != NULL) 
	{
		/* Some more variables */

		char line[1000];
		char *part;
		int q = 0, delV = 0, vBatCode, vBatRegValue, totalDelay = 0;
		
		/* read a line from file */

		while(fgets(line, sizeof line, file1) != NULL)		
		{
			printf("%s", line);  // print the line contents to stdout

			// parse the line to get time and voltage
			part = strtok(line, ",");
			while(part != NULL) 
			{
				// convert the extracted part to float	
				parts[q] = atof(part);
				// continue parsing
				part = strtok(NULL, ",");
				q++;	
			}
			// reset the array counter for next line
			q = 0;	
			
			printf("Time: %lf, voltage: %lf \n", parts[0], parts[1]);
			
			/* Logic for converting voltage to code as needed by the charger */

			// Get voltage over 3.5V, in milivolts
			delV = parts[1] * 1000 - 3500;
			
			// Regularize bad voltages, no voltage allowed under 3.5 V and over 4.44 V
			if (delV < 0)
			{
				delV = 0;
			}
			else if (delV > 900)
			{
				delV = 900;
			}
			
			// Our resolution is 20 mV
			vBatCode = delV / 20;
			// as the first two bits are zero, pg - 33  of datasheet
			vBatRegValue = vBatCode * 4;	
			
			// The zero problem, the voltage starts with 3.6, even trying to set it to 3.5, so rather, starting with 3.52
			if(vBatRegValue == 0)
			{
				data2[1] = 4; //vBatRegValue;
			}
			else 
			{
				data2[1] = vBatRegValue;
			}
			// This is the time, from the text file.	
			totalDelay = parts[0];
			
			setRegisters: 				// GOTO should be used **very rarely**	
			
			// Find the time of register write
			time (&rawtime);
			timeinfo = localtime(&rawtime);
			printf("Current local time and data: %s \n", asctime(timeinfo));
			printf("Setting voltage to: %lf \n", 3500.0 + delV);

			// Write data to the I2C device, one register at a time
			printf("data0[1] : %d, data1[1]: %d, data2[1]: %d, data3[1]: %d \n", data0[1], data1[1], data2[1], data3[1]);
			printf("data4[1] : %d, data5[1]: %d, data6[1]: %d \n", data4[1], data5[1], data6[1]);
			I2cSendData(BQ24261_ADDR, data0, 2);
			I2cSendData(BQ24261_ADDR, data1, 2);
			I2cSendData(BQ24261_ADDR, data2, 2);
			I2cSendData(BQ24261_ADDR, data3, 2);
			I2cSendData(BQ24261_ADDR, data4, 2);
			I2cSendData(BQ24261_ADDR, data5, 2);
			I2cSendData(BQ24261_ADDR, data6, 2);
			
			/* Timing Logic */	
			
			// The voltage value needs to be re-written to the charger every few seconds (10 here)	
			while (totalDelay > 0) 
			{	
				// using delay, makes this platform specific
				if (totalDelay > 10) 
				{
					totalDelay = totalDelay - 10;
					delay(10000);
					goto setRegisters;
				}
				else if (totalDelay <= 10) 
				{
					delay(totalDelay * 1000);					
					totalDelay = 0;
				}	
					
			}

		}
		
		// Close the file	
		fclose(file1);
		printf("\n\n\n *****Charging done !!!*****");
	}
	else
	{
		perror(filename1);
	}
	
	printf("\n\n\n *****Charging done !!!*****");	
        close(deviceDescriptor);

        endwin();

        return 0;
}

File: BQ-24261-Charger/js/charger.js
// This is the js file to interact with the charger

var i2c = require('i2c');
var LineByLineReader = require('line-by-line');
var delayed = require('delayed');

var address = 0x6B;  
var inputFile = "testdata.txt";

var lr = new LineByLineReader(inputFile);

var device1 = new i2c(address, {device: '/dev/i2c-1', debug: false});

var data0 = [];
var data1 = [];
var data2 = [];
var data3 = [];
var data4 = [];
var data5 = [];
var data6 = [];

data0[0] = 0;
data0[1] = 148; // 16; //144;   // make the voltage permanent
data1[0] = 1;
data1[1] = 64;
data2[0] = 2;
data2[1] = 4;
data3[0] = 3;
data3[1] = 70;
data4[0] = 4;
data4[1] = 42;
data5[0] = 5;
data5[1] = 0;
data6[0] = 6;
data6[1] = 112;

lr.on('error', function (err) {
    // 'err' contains error object
    if(err)
        console.log("Error opening input file : " + err);
});

lr.on('line', function (line) {
    
    lr.pause();
    
    // 'line' contains the current line without the trailing newline character.
    console.log("Line read : " + line);
    
    // Each line is an ordered-pair of time,current,voltage
    // Only one of current or voltage, will be non-zero.
    var lineParts = line.toString().split(",");
    var time = parseFloat(lineParts[0]);
    var current = parseFloat(lineParts[1]);
    var voltage = parseFloat(lineParts[2]);
    
    if (current == 0) {
        console.log("Trying to set voltage to : " + voltage + " V for a time of : " + time);
        
         /* Logic for converting voltage to code as needed by the charger */

    	// Get voltage over 3.5V, in milivolts
    	var delV = voltage * 1000 - 3500;
    	
    	// Regularize bad voltages, no voltage allowed under 3.5 V and over 4.44 V
    	if (delV < 0) {
    		
    		delV = 0;
    	}
    	else if (delV > 900) {
    		
    		delV = 900;
    	}
    	
    	// Our resolution is 20 mV
    	var vBatCode = delV / 20;
    	// as the first two bits are zero, pg - 33  of datasheet
    	var vBatRegValue = vBatCode * 4;	
    	
    	// The zero problem, the voltage starts with 3.6, even trying to set it to 3.5, so rather, starting with 3.52
    	if(vBatRegValue < 4) {
    		
    		data2[1] = 4; //vBatRegValue;
    	}
    	else {
    	    
    		data2[1] = vBatRegValue;
    	}
        
        var timeDelay = time;
        
        setValues (data0, data1, data2, data3, data4, data5, data6);
        
        delayed.delay( function() {
            
            lr.resume();
        
        }, timeDelay*1000);
        
        while (timeDelay > 0.0) {
            
            if (timeDelay > 10.0) {
    
                delayed.delay( function() {
                    
                    setValues (data0, data1, data2, data3, data4, data5, data6);
                    console.log("TimeDelay :" + timeDelay);
    
                }, timeDelay*1000);
    
                timeDelay = timeDelay - 10.0;
    
            }
            
            else if (timeDelay <= 10) {
            
                delayed.delay( function() { 
    
                }, timeDelay*1000);
    
                timeDelay = 0.0;
    
            }
        }
    } else if (voltage == 0) {
        console.log("Trying to set current to : " + current + " mA for a time of : " + time);
        
         /* Logic for converting current to code as needed by the charger */

    	// Get current over 500 mA
    	var delI = current - 500;
    	
    	// Regularize bad currents, no current allowed under 500 mA and over 3000 mA
    	if (delI < 0) {
    	
    		delI = 0;
    	} else if (delI > 2500) {
    		
    		delI = 2500;
    	}
    	
    	// Our resolution is 100 mA
    	var iBatCode = delI / 100;
    	// as the first three bits are 010, pg - 33  of datasheet
    	var iBatRegValue = iBatCode * 8 + 2;	
    	
    	data2[1] = iBatRegValue;
    	/*// The zero problem, the voltage starts with 3.6, even trying to set it to 3.5, so rather, starting with 3.52
    	if(vBatRegValue < 4) {
    		
    		data2[1] = 4; //vBatRegValue;
    	}
    	else {
    	    
    		data2[1] = vBatRegValue;
    	}*/
    	
        
        var timeDelay = time;
        
        setValues (data0, data1, data2, data3, data4, data5, data6);
        
        delayed.delay( function() {
            
            lr.resume();
        
        }, timeDelay*1000);
        
        while (timeDelay > 0.0) {
            
            if (timeDelay > 10.0) {
    
                delayed.delay( function() {
                    
                    setValues (data0, data1, data2, data3, data4, data5, data6);
                    console.log("TimeDelay :" + timeDelay);
    
                }, timeDelay*1000);
    
                timeDelay = timeDelay - 10.0;
    
            }
            
            else if (timeDelay <= 10) {
            
                delayed.delay( function() { 
    
                }, timeDelay*1000);
    
                timeDelay = 0.0;
    
            }
        }
    } else {
        console.log("Something wrong with the input, eiher one of voltage or current should be zero.");
    }
});

lr.on('end', function () {
    
    // All lines are read, file is closed now.
    console.log("End of input file");
    console.log("*****Charging done !!!****");

});

function setValues (data0, data1, data2, data3, data4, data5, data6) {
    
    console.log("Setting values at "  + getDateTime());
    
    device1.write(data0, function(err) { if(err) console.log("error in writing data0[0]"); });
    device1.write(data1, function(err) { if(err) console.log("error in writing data1[0]"); });
    device1.write(data2, function(err) { if(err) console.log("error in writing data2[0]"); });
    device1.write(data3, function(err) { if(err) console.log("error in writing data3[0]"); });   
    device1.write(data4, function(err) { if(err) console.log("error in writing data4[0]"); });
    device1.write(data5, function(err) { if(err) console.log("error in writing data5[0]"); });
    device1.write(data6, function(err) { if(err) console.log("error in writing data6[0]"); });
    
}

function getDateTime() {

    var date = new Date();

    var hour = date.getHours();
    hour = (hour < 10 ? "0" : "") + hour;

    var min  = date.getMinutes();
    min = (min < 10 ? "0" : "") + min;

    var sec  = date.getSeconds();
    sec = (sec < 10 ? "0" : "") + sec;

    var year = date.getFullYear();

    var month = date.getMonth() + 1;
    month = (month < 10 ? "0" : "") + month;

    var day  = date.getDate();
    day = (day < 10 ? "0" : "") + day;

    return year + ":" + month + ":" + day + ":" + hour + ":" + min + ":" + sec;

}


